### ISSUE TITLE
Add facilities to support story structure analysis and developmental edits

### ISSUE THREAD

**Issue by instancezero** [vkbo/novelWriter#1784]
This idea comes as a result of the discussion at https://github.com/vkbo/novelWriter/discussions/1769

Add a new "Story" construct of the form `%Story[.<term>: <text>` to NovelWriter, where auto-complete can suggest any of a set of predefined or user-defined values for `<term>`, and where `<text>` is nominally free-form but MAY provide some validation rules.

Collate the story structure metadata in the preview view, giving users the ability to show/hide the data.

Integrate this with existing NovelWriter metadata commands to provide a facility for exporting scene-level summaries of the story structure in formats that are accessible to a wide range of users and applications (CSV for spreadsheets and databases, JSON for general use, and possibly ODS).

**Comment by vkbo** [comment]
Thanks for writing this up and for the discussion. A few implementation notes for future reference:

* The `term` part is free form, and should be collected in the index.
* Auto-completion can help so the user don't have to memorise or look up all their defined terms.
* Story and synopsis comments, and other prefixed comments, should be collected in a separate new structure in the index.
* How these comments show up in the Outline needs to be figured out.
* They can easily be added to the Outline export feature.
* Adding JSON should be easy enough.

Also:

* ODS is maybe a little redundant in addition to CSV. Especially considering how much work it is to write ODF export code. New dependencies will not be pulled in.

**Index example:**

```json
"88706ddc78b1b": {
  "headings": {
    "T0001": {"level": "H2", "title": "Where has John Gone?", "line": 1, "tag": "", "cCount": 20, "wCount": 4, "pCount": 0},
    "T0002": {"level": "H3", "title": "Jane Cannot Find John", "line": 8, "tag": "", "cCount": 119, "wCount": 24, "pCount": 1}
  },
  "references": {
    "T0001": {"jane": "@pov", "space": "@location"},
    "T0002": {"jane": "@pov", "john": "@focus", "space": "@location"}
  },
  "comments": {
    "synopsis": "text",
    "story.term": "text",
  }
}
```

**Comment by xahodo** [comment]
This would be really helpful to me. I'm currently in the process of abusing `%Synopsis:` for this.

For me it would be helpful to put some data points right in the chapter/scene, which I could later reference in an external program.

Some things I could use this for:
* Conflict (what makes the scene exciting)
* Goal (what I want to reach with the scene)
* Event (significant story event happening in the scene)

Yes, these might be satisfied by the `@plot` tag reference, but I like to have them right in the scene (for when I'm actually writing).

Writing this down in a format like `%Story: <Conflict>John has to choose between his fiancé and his mistress.` Why the angled brackets? They are more discernible.

If NW would have an interface to view these nicely organized in one screen it would make this feature even more fantastic. Things would become awesome if you could filter based on which terms to display (in case of there being several).

**Comment by vkbo** [comment]
The syntax is already locked down and implemented in the code as it is used for footnotes for the 2.5 release. So it will be:

```md
### Scene

%Story.Conflict: John has to choose between his fiancé and his mistress.

Bla bla bla.
```

Anyway, I've scheduled this for inclusion in 2.6 as the 2.5 release is already very full and I don't want to add too much new notation in one release. The footnotes were just higher on the list.

Adding this feature will include an update to the Outline View in some form yet to be decided. I think it needs a little trial and error, so I would be happy to have some back and forth with some of the people interested in this feature to sort out the details. @instancezero has made a POC implementation in a branch.

**Comment by vkbo** [comment]
I'm adding a task list here to track progress.

* [x] Update the index to support the data structure needed. Implementation: #2255.
* [ ] Add tokenizer and syntax highlighter parsing of story structure comments. Implementation: to do.
* [ ] Add a model view for the outline to simplify addition of data columns. Implementation: to do.
* [ ] Update the outline panel to show story structure info. Implementation: to do.
* [ ] Extend export options for story structure data. Implementation: to do.

**Comment by vkbo** [comment]
The 2.7 release is already getting really big, so I think I'm going to split this task between 2.7 and 2.8.

For 2.7 will include the restructuring of the index, which is already done in #2255. I will also enable the syntax highlighting for story keys (the code is already there as it is also used for footnotes) and the parsing of these comments. This is mostly done already in #2284, but I still need to format them for manuscript output. I will also add support for exporting them from the Outline panel.

However, I will postpone everything else related to displaying them on the Outline until 2.8. Mainly because I am working on a completely new Outline view that will incorporate some of the timeline features too, like #76 and #854. I do not want to have to add the story feature twice to the Outline, but waiting to add the timeline features first is going to take too long.

Point 4 and 5 on the task list will therefore be moved to 2.8 as new tasks and removed from here.

**Comment by instancezero** [comment]
> The 2.7 release is already getting really big, so I think I'm going to split this task between 2.7 and 2.8.
> 

Makes sense to me. I haven't pushed the code yet, mostly for a lack of documentation, but my "export to ODS" tool now supports customizing what gets extracted, some filtering (so I can assign a column for each of my main characters and stuff the rest into an "others" column, and bolding the first appearance of a character, location, etc. It has proved invaluable as I have been adding a sub-plot and dealing with some extensive "refactoring" that is required to make everything work.

**Comment by vkbo** [comment]
I've added the necessary code to collect all story structure keys/terms used in a unique set so they can be used for an auto-complete menu. The keys will be case sensitive though. I don't see a way around that since they are not defined in a singular place like tags are. It's sort of the same problem that the auto-complete of post tags on Mastodon has.

**Comment by vkbo** [comment]
> I haven't pushed the code yet, mostly for a lack of documentation, but my "export to ODS" tool now supports customizing what gets extracted, some filtering (so I can assign a column for each of my main characters and stuff the rest into an "others" column, and bolding the first appearance of a character, location, etc. It has proved invaluable as I have been adding a sub-plot and dealing with some extensive "refactoring" that is required to make everything work.

On a side note, I'm considering writing a story structure and time line tool as an external app. Partially because I may use C++ for this instead. It frees me up to do a lot more real time graphics rendering than I can from Python due to performance issues.

I think I would write it to parse a dedicated JSON export file that novelWriter could produce. I also have a C++/Qt6 writing app I'm tinkering with, so a standalone app for the analysis would be useable by both.

**Comment by instancezero** [comment]
I'd be happy to offer a brief write-up, although the real value in the method is to pull it out to a chart to look at overall structure.

I'm in the middle of documenting the update to NovelWriterExtract and have decided that I really need a test suite before releasing it, which has me writing code again instead of a story. Duh + LOL.

Here's a sample of what the extract tool is producing as I add two sub-plots to the story (synopsis not shown so I don't give too much away):

![Image](https://github.com/user-attachments/assets/770352ca-c32b-4f1d-9f0e-86583f9868ea)

The first occurrence of a location, plot line (P1, P2, P3) or my main characters ("Shawn" in the story is compressed to "Sh" to keep things compact), then secondary characters, the scene word count, and then the synopsis. At the moment, I'm using this in a way similar the @mention construct, to make sure a character is not introduced after they've shown up in a scene. Once this draft is complete, I'll add/update the story tags, do an extract of that information and look at things like pacing.

**Comment by instancezero** [comment]
Here's a run at it!

# Story Structure in novelWriter

Version x.y of novelWriter supports story structure using the %story style of comment.
To use the feature, make the first word of a comment `story`, followed by a period, 
a structure term, a colon, a space and the text for that term.

```
%story.term: text
```

The term can be anything that you want to track in the manuscript.
This construct is intended to make it easier to extract metadata from a work to perform a structural analysis of the story.

There are probably as many ways to examine story structure as there are authors and editors combined.
For this reason the story tag is flexible.
You can use any terms you want and track any aspect of the story that serves your purposes.
An example method has been advanced by Shawn Coyne in _The Story Grid_.
This method asserts that a story is composed of "beats", and that each beat has an inciting incident, a complication, a crisis, and a resolution.
One might capture these elements of a beat where a character overcomes their fear of giving a speech as:

```
%Synopsis: Carol overcomes her fear of giving a speech.
%Story.incite: Carol is pleased to be invited to a confernece to see her boss deliver a keynote.
%Story.complication: Carol's boss calls in sick and asks her to deliver a big speech.
%Story.crisis: Carol has a fear of appearing on stage.
%Story.resolution: Carol engages the help of a coach who helps her overcome her fears and delivers a great speech.
```

Other analyical models propose tracking a scene's pace, how it affects the mood of the story,
or which element(s) of the story's genre are being satisfied.
An author can use this mechanism to track any element of a scene.
Some examples include time of day, how much time passes in the scene, ot even the physical form of a shape-shifting character.
If a story involves magic, one could track which wand a main characteh has in hand. It's up to tha author.

When the story and other scene metadata is extracted into a tabluar form,
it is possible to get a comprehensive overview of the story and to identify possible issues
(for example, so many fast-paced scenes without a break that readers might become fatigued or over-stimulated).

**Comment by vkbo** [comment]
That's a good start. Could you submit it as a pull request? Create a file `docs/source/project_story.rst` and copy/paste the markdown there. Only a few things need to be changed to work with rst files. I can plug it into the rest of the docs.

The title needs to be:

```rst
***************
Story Structure
***************
```

The example needs to be in a block:

```rst
.. code-block:: md

   %Synopsis: Carol overcomes her fear of giving a speech.
```

I prefer contributions to be from the person who contributed them, but I can do it too of course if needed.

**Comment by instancezero** [comment]
Will do. I'm slammed for the next few days so probably Monday before I can get to it.

**Comment by vkbo** [comment]
The documentation isn't published until the final release, which is still at least a month away.

I have a bit of a deadline for 2.7 on the account of switching from Qt5 to Qt6, which is a bit of an exception to my usual "I will release it when it's ready" approach. 😄

Qt5 reaches end of life on 26 May, so I would like to release 2.7 before that, which means I need to make the beta release within a couple of weeks. That's also why I'm postponing the Outline View updates until 2.8.

I merged the parsing and processing of story structure comments yesterday though, so it's in. And I will add the CSV export soon. That shouldn't be very tricky anyway. I'll see about the auto-complete menu. I may not add it until the RC1 release.
